src/lib/Endpoint/SqlSrvUsers.php

```diff
@@ -0,0 +1,690 @@
+            $this->logger->error('failed to count number of objects from endpoint', [
+                'class' => get_class($this),
+                'exception' => $e,
+            ]);
+
+            return 0;
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getAll(?array $query = null): Generator
+    {
+        list($filter, $values) = $this->transformQuery($query);
+        $this->logGetAll($filter);
+
+        if ($filter === null) {
+            $sql = self::USERQUERY;
+        } else {
+            $sql = self::USERQUERY.' WHERE '.$filter;
+        }
+
+        try {
+            $result = $this->socket->prepareValues($sql, $values);
+        } catch (InvalidQuery $e) {
+            $this->logger->error('failed to fetch resources from endpoint', [
+                'class' => get_class($this),
+                'exception' => $e,
+            ]);
+
+            return 0;
+        }
+
+        $i = 0;
+        $result = $this->socket->getQueryResult($result);
+
+        foreach ($result as $object) {
+            yield $this->build($this->prepareRawObject($object));
+            ++$i;
+        }
+
+        return $i;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getOne(array $object, array $attributes = []): EndpointObjectInterface
+    {
+        list($filter, $values) = $query = $this->transformQuery($this->getFilterOne($object));
+        $this->logGetOne($filter);
+
+        $sql = self::USERQUERY.' WHERE '.$filter;
+        $result = $this->socket->prepareValues($sql, $values);
+        $result = $this->socket->getQueryResult($result);
+
+        if (count($result) > 1) {
+            throw new Exception\ObjectMultipleFound('found more than one object with filter '.$filter);
+        }
+        $return = array_shift($result);
+        if ($return === null) {
+            throw new Exception\ObjectNotFound('no object found with filter '.$filter);
+        }
+
+        return $this->build($this->prepareRawObject($return), $query);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function create(AttributeMapInterface $map, array $object, bool $simulate = false): ?string
+    {
+        $this->logCreate($object);
+        $loginName = $this->getNameByAttribute($object, self::ATTRLOGINNAME);
+
+        if ($simulate === true) {
+            return null;
+        }
+
+        try {
+            if (isset($object['mechanism']) && $object['mechanism'] === 'windows') {
+                $query = $this->createWindowsLogin($loginName);
+            } else {
+                $query = $this->createLocalLogin($loginName, $object['password'], $object[self::ATTRHASTOCHANGEPWD] ?? true);
+            }
+
+            $this->socket->query($query, $simulate);
+
+            if (isset($object['disabled']) && $object['disabled'] === true) {
+                $this->disableLogin($loginName, $simulate);
+            }
+
+            $principalId = $this->getPrincipalId($loginName);
```

Comment:
You should validate if principal is valid since it has to be fetched with an additional query and is not a result from the previous operation. And if not throw an exception. Might be even better to check it in `getPrincipalId()`.